base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -643,17 +652,54 @@ public void validateJWS(JWS jws, String alg, JWK jwk) throws Exception {
 
             RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, publicExponent);
             publicKey = keyFactory.generatePublic(keySpec);
+        } else if ("ES256".equals(alg)){
+            signer = Signature.getInstance("SHA256withECDSA", "Mozilla-JSS");
 
+            String kty = jwk.getKty();
+            KeyFactory keyFactory = KeyFactory.getInstance(kty, "Mozilla-JSS");
+            AlgorithmParameters algoParameters = AlgorithmParameters.getInstance(kty);
+
+            String crv = jwk.getCrv();
+            ECGenParameterSpec ecGenParameterSpec;
+            if ("P-256".equals(crv)) {
+                ecGenParameterSpec = new ECGenParameterSpec("secp256r1");
+            } else {
+                ACMEError error = new ACMEError();
+                error.setType("urn:ietf:params:acme:error:badSignatureAlgorithm");
+                error.setDetail("EC curve of type " + crv + " not supported\n" +
+                                "Try again with P-256.");
+                throw new ACMEException(HttpServletResponse.SC_BAD_REQUEST, error);
+            }
+            algoParameters.init(ecGenParameterSpec);
+            ECParameterSpec ecParameterSpec = algoParameters.getParameterSpec(ECParameterSpec.class);
+
+            String x = jwk.getX();
+            BigInteger biX = new BigInteger(1, Base64.decodeBase64(x));
+
+            String y = jwk.getY();
+            BigInteger biY = new BigInteger(1, Base64.decodeBase64(y));
+
+            ECPoint ecPoint = new ECPoint(biX, biY);
+            ECPublicKeySpec ecKeySpec = new ECPublicKeySpec(ecPoint, ecParameterSpec);
+            publicKey = keyFactory.generatePublic(ecKeySpec);
+            logger.debug("PublicKey: {}", Base64.encodeBase64String(publicKey.getEncoded()));
+            byte[] rawSign = Base64.decodeBase64(jws.getSignature());
+            
+            SEQUENCE seq = new SEQUENCE();
+            INTEGER i1 = new INTEGER(Arrays.copyOfRange(rawSign, 0, 32));
+            INTEGER i2 = new INTEGER(Arrays.copyOfRange(rawSign, 32, 64));
+
+            seq.addElement(i1);
+            seq.addElement(i2);
+            jws.setSignature(Base64.encodeBase64URLSafeString(ASN1Util.encode(seq)));
```

Comment:
Could you add inline comments to describe why this code is needed?
Does this mean caddy is generating a non-standard ES256 signature?
Or does it mean NSS/JSS is not able to validate a standard ES256 signature?

Reply:
If they are indeed 2 different formats, I'd suggest to keep the signatures in separate variables, probably something like this:
```
byte[] jssSignature;
if ("RS256".equals(alg)) {
    ...
    jssSignature = Base64.decodeBase64(jws.getSignature());
} else if ("ES256".equals(alg)) {
    ...
    byte[] jwsSignature = Base64.decodeBase64(jws.getSignature());
    ... convert signature ...
    jssSignature = ASN1Util.encode(sequence);
}

validateJWS(signer, jws, jssSignature);
```

Reply:
BTW, the two `validateJWS()` methods probably can just be merged.